### PR TITLE
8260337: Optimize ImageReader lookup, used by Class.getResource

### PR DESCRIPTION
**Pull request by cl4es**
This patch optimizes the code paths exercised by `String.class.getResource("String.class")` by:

- Adding an ASCII fast-path to methods verifying strings in the jimage, which can then be done allocation-free
- Avoiding the allocation of the `long[8]` attributes when verifying only for the purpose of verifying a path exists
- Using the `JNUA.create` fast-path in `SystemModuleReader` (which should be OK since we just verified the given name is a JRT path)
- Remove a redundant check in `Class::resolveName` and fitting the `StringBuilder` to size

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8260337](https://bugs.openjdk.java.net/browse/JDK-8260337): Optimize ImageReader lookup, used by Class.getResource

### Reviewers
 * [Jim Laskey](https://openjdk.java.net/census#jlaskey) (@JimLaskey - **Reviewer**)
 * [Athijegannathan Sundararajan](https://openjdk.java.net/census#sundar) (@sundararajana - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/2212/head:pull/2212`
`$ git checkout pull/2212`

### CODE REVIEW

**Comment by cl4es** [comment]
On the provided benchmark we can observe a 2x speed-up and a 70% reduction in allocation.

Baseline:
```
Benchmark                                                      Mode  Cnt     Score     Error   Units
ClassGetResource.stringClass                                   avgt    5  2885.403 ± 309.787   ns/op
ClassGetResource.stringClass:·gc.alloc.rate.norm               avgt    5  1368.124 ±   0.056    B/op
```

Patch:
```
Benchmark                                                      Mode  Cnt     Score     Error   Units
ClassGetResource.stringClass                                   avgt    5  1411.980 ± 155.877   ns/op
ClassGetResource.stringClass:·gc.alloc.rate.norm               avgt    5   408.038 ±   0.043    B/op
```

File: src/java.base/share/classes/jdk/internal/module/SystemModuleFinders.java [hunk 1]
```diff
@@ -432,11 +432,27 @@ private ImageLocation findImageLocation(String name) throws IOException {
             }
         }
 
+        /**
+         * Verifies the existence of the given resource, {@code false}
+         * if not found.
+         */
+        private boolean containsLocation(String name) throws IOException {
```
Comment (AlanBateman) [inline src/java.base/share/classes/jdk/internal/module/SystemModuleFinders.java:439]: Can you rename this to containsImageLocation to keep it consistent with findImageLocation? Alternative rename findImageLocation. Also would be better for the description to be "Returns true if the given resource exists, false if not found".

The changes to the jimage code will take time to review, probably should have 2 reviewers.

Comment (cl4es) [reply src/java.base/share/classes/jdk/internal/module/SystemModuleFinders.java]: Done. 

I've looped in @JimLaskey

**Review by JimLaskey — APPROVED** [review]
Nice clean up. LGTM

**Comment by AlanBateman** [comment]
@JimLaskey Do you have anything that documents the jimage format that could be checked into the repo to help with future changes in this change? Clearly any such document would need a warning in 96pt font that the format is JDK internal, submit to change from build to build, etc. but it could be helpful when working on this code.

**Comment by cl4es** [comment]
@sundararajana @AlanBateman - can I solicit a second review on this? Or an OK from Alan to go ahead and push with the one review.

**Review by sundararajana — APPROVED** [review]
LGTM

**Comment by cl4es** [comment]
@JimLaskey @sundararajana - thanks for reviewing!

/integrate

**Pull request closed without merging**